climada/engine/impact.py

```diff
@@ -1409,123 +1417,177 @@ def concat(cls, imp_list,concat_type='time_window'):
+
+        #if impact list is empty, return cls()
         if len(imp_list) == 0:
             return cls()
-       
-        if concat_type not in ['time_window','exposure']:
+
+        #raise error if concat_type is not available
+        #NOTE: not sure how to use an ENUM class in this context
+        if concat_type not in ['time','exposure']:
             raise ValueError(f"Concatenation option {concat_type} not available.")
-       
-        #initialize new impact object
-       
+
         # check impacts consistency and fill impact object
-       
-        ### the following checks are common for both concat options
-       
+
+        ### the following checks are in common for both concat options
         # check hazard type
-        haz_types = {imp.tag['haz'].haz_type for imp in imp_list if imp.tag['haz'].haz_type != ''}
-        if len(haz_types) > 1:
-            raise ValueError(f"The impacts are based on hazards of different types: {haz_types}. "
-                            "The impacts are incompatible and cannot be concatenated.")      
-       
+        check_haz_types(imp_list)
         # check unit
-        units = {imp.unit for imp in imp_list if imp.unit != ''}
-        if len(units) > 1:
-            raise ValueError(f"The given impacts have different units: {units}. "
-                            "The impacts are incompatible and cannot be concatenated.")
-        if len(units) == 0:
-            units = {''}
-        unit = units.pop()
-        
+        unit = check_units(imp_list)
         #check coordinate reference system
-        crss={imp.crs for imp in imp_list if imp.crs != ''}
-        if len(crss)>1:
-            raise ValueError(f"The given impacts have different crs: {crss}. "
-                        "The impacts are incompatible and cannot be concatenated.")
-        crs = crss.pop()
-        
+        crs = check_crs(imp_list)
         #check attributes
-        attributes = sorted(set.union(*[set(vars(imp).keys()) for imp in imp_list]))
-        for attr_name in attributes:
-            if not all(hasattr(imp, attr_name) for imp in imp_list):
-               raise ValueError(f"Attribute {attr_name} is not shared by all impacts. "
-                                "The impacts are incompatible and cannot be concatenated.")
-
+        check_attributes(imp_list)
         #check if impacts have same frequency values (only first entries are checked)
-        freqs=set([imp.frequency[0] for imp in imp_list])
-        if len(freqs)>1:
-            raise ValueError(f"Impacts do not have the same frequency: {freqs}."
-                            "The impacts are incompatible and cannot be concatenated.")
-        frequency = np.concatenate([imp.frequency for imp in imp_list], axis=0)
-       
+        #unclear if this check can be avoided to allow concatenating impacts
+        #with different frequencies
+        #but this affects other attributes such as eai
+        frequency = check_frequency(imp_list)
```

Comment:
This should only be done if the impact matrix is not defined. Otherwise, there is no need for identical frequencies.

Reply:
ok, you are right. I remove the frequency check.